### DISCUSSION TITLE
Using a Repeating Panel Items as an input

### DISCUSSION THREAD

**Discussion by lewysigns** [anvilistas/zod-validator#2] [Q&A]
I was reviewing the input_helpers.py file to see what I could use to automate my validation and notice repeating panels is not a form of input.

I am attempting to add it here (Not sure if the test branch will be in the clone) :

https://anvil.works/build#clone:6MCKBCN52M4PVMXY=I3MIZPOCFAPXV4HNTKY2YTMZ

I added it to the input_helpers.py:

```python
_value_property_map = {
    anvil.TextBox: _prop("text"),
    anvil.DatePicker: _prop("date"),
    anvil.DropDown: _prop("selected_value"),
    anvil.RepeatingPanel: _prop("items")
}
```

Ideally this would allow me to add a schema check like so:

```python
user_schema = z.typed_dict(
    {
        "name": z.string().strip().min(3, message="Name must be atleast 3 letters").max(50).transform(str.capitalize).refine(lambda s: s != "Admin", message="You can't be an admin"),
        "email": z.string().strip().email(),
        "age": z.coerce.integer(invalid_type_error="Expected a valid age").ge(18).lt(100),
        "dob": z.date().max(date.today() - timedelta(days=365*18), "Must be older than 18"),
        "color": z.string().upper().optional().pipe(z.enum(["RED", "GREEN", "BLUE"], invalid_type_error="Must be 'red', 'green' or 'blue'")),
        "todos":z.typed_dict({
          'name':z.string().min(2,message="Name needs to be atleast 2 characters"),
          "detail":z.string().min(3,message="detail needs to be atleast 3 characters")
        }).list() # This is the validation I am trying to acheive
    }
```

I know I could add an individual schema to the repeating panel row, but how would that get aggregated and checked by the over all validator? Or would I just need set every row's validator to the same submit button?

There I get is in the Validtor.py trying to set the event of the repeating_panel (It does not have the change event). This issue is there isn't exactly an event that exist when the items update at the repeating panel level. Would I need to create a custom event thats fired when a repeating panel row's data bindings are refreshed?

```python 
#failing update in Validator.py __init__

        for key in schema.shape:
            self.inputs[key] = getattr(form, key + input_suffix)
            self.errors[key] = getattr(form, key + error_suffix)
            on_change = self.change_handler(key)
            try:
              self.inputs[key].add_event_handler("change", on_change)
            except:
              try:
                self.inputs[key].add_event_handler("refreshing_data_bindings", on_change)
              except Exception as e:
                raise e
```

**Comment by s-cork** [comment]
you'll have to make it the master branch and ping me when you've done that so I can re-clone it.
(cloning currently only ever clones the master branch)

**Reply by lewysigns** [reply]
Here it is!

It's not quite working..

https://anvil.works/build#clone:ONDDNCWCEJDAYZIN=MNG6ZDVJXZMKWPLSQ6NCN4KD

**Reply by s-cork** [reply]
Great - it's almost working it's just that the update error isn't quite doing what you want.
The path to the error is `["todos", 0, "name']` for example.

e.g.

```python
>>> x = z.typed_dict(
    {
        "todos":z.typed_dict({
          'name':z.string().min(2,message="Name needs to be atleast 2 characters"),
          "detail":z.string().min(3,message="detail needs to be atleast 3 characters")
        }).list()
    }
)
>>> result = x.safe_parse({"todos": [{"name": " "}]})
>>> result
ParseResult(success=False, error=Name needs to be atleast 2 characters at ['todos'][0]['name']; Required at ['todos'][0]['detail'])
>>> result.error
ParseError("Name needs to be atleast 2 characters at ['todos'][0]['name']; Required at ['todos'][0]['detail']")
>>> result.error.errors(["todos"])
[]
>>> result.error.errors(["todos", 0, "name"])
['Name needs to be atleast 2 characters']
```

**Reply by lewysigns** [reply]
I see.

I think the least path of least resistance is to set up a validator for each row. 

Though, I am trying to update my version to handle repeating panels by setting up some unique input-error for each row added if we see the input type is an anvil.RepeatingPanel instance. It just gets a little sticky because the number of rows isn't set at the time the form is initiated.

**Reply by s-cork** [reply]
Yeah I don't know what the right approach is here. I feel like there needs to be a way to get the error paths from the error object. 

And then for each error path you could some how walk the path to get to the component. 

But that would require the RepeatingPanel item template to know its index. Which is possible. 

Another way might be to treat extract the zod element from the zod list. And make each ItemTemplate a validator in its own right. 

`todo_schema = user_schama.shape['todos'].element()`

I think. 

Let us know what you come up with. 

I might look at some other zod helpers and see if they have any useful approaches.
